File: java/src/main/java/fizz/buzz/whizz/Matcher.java
package fizz.buzz.whizz;

import static java.lang.String.valueOf;

@FunctionalInterface
public interface Matcher {
  boolean matches(int n);

  static Matcher times(int times) {
    return n -> n % times == 0;
  }

  static Matcher contains(int num) {
    return n -> valueOf(n).contains(valueOf(num));
  }

  static Matcher always(boolean bool) {
    return n -> bool;
  }
}

File: scala/src/main/scala/fizz/buzz/whizz/Matcher.scala
package fizz.buzz.whizz

object Matcher {
  type Matcher = Int => Boolean

  def times(n: Int): Matcher = _ % n == 0
  def contains(n: Int): Matcher = _.toString.contains(n.toString)
  def always(bool: Boolean): Matcher = _ => bool
}
